### PR TITLE
BUG: "sp.linalg.solve_discrete_are" fails for random data

### LINKED ISSUE #2251: solve_discrete_are in scipy.linalg does (sometimes) not solve correctly (Trac #1732)
_Original ticket http://projects.scipy.org/scipy/ticket/1732 on 2012-09-20 by trac user inde, assigned to unknown._

Hi,
this is for scipy 0.11.0rc2. It seemed to be a normal release on sourceforge, but there is no 0.11 here, so I am filing under devel.

So the problem I am observing when solving the discrete algebraic Riccati Equation for the linear quadratic regulator problem is that sometimes the solution returned by scipy.linalg.solve_discrete_are is not positive definite or even correct.

I am testing with 3x3 random matrices (code below). The attached npz file contains one particular case that did not solve for me.

    def dare_zero_compare(self,A,B,Q,R,S):
        AT_S_A = s.dot(A.T,s.dot(S,A))
        AT_S_B = s.dot(A.T,s.dot(S,B))
        BT_S_B_plusR_inv = s.linalg.inv(s.dot(B.T,s.dot(S,B)) + R)
        BT_S_A = s.dot(B.T,s.dot(S,A))

        zeroS = AT_S_A - S - s.dot(AT_S_B, s.dot(BT_S_B_plusR_inv, BT_S_A)) + Q
        return zeroS

    def test_dareSsolves(self):
        for i in range(100):
            n = 3
            A = s.random.rand(n,n)
            B = s.random.rand(n,1)
            R = s.eye(B.shape[1])
            Q = s.eye(A.shape[0])

            
            S = s.linalg.solve_discrete_are(A,B,Q,R)
            
            zeroS = self.dare_zero_compare(A,B,Q,R,S)            
            try:
                for ai in zeroS:
                    for aij in ai:
                        self.assertAlmostEqual(aij, 0.0)
            except AssertionError:
                with open ('test_dareSsolves.m','w') as f:
                    f.write(string_dare_test(A,B,Q,R,S) )
                s.savez('test_dareSsolves.npz', A=A,B=B,Q=Q,R=R,S=S)

    def test_dareSsolves_repeat(self):
        npzfile = s.load('test_dareSsolves.npz')
        A = npzfile['A']
        B = npzfile['B']
        Q = npzfile['Q']
        R = npzfile['R']
        S = npzfile['S']
                
        S_new = s.linalg.solve_discrete_are(A,B,Q,R)
        print A,B, Q,R,S,S_new
        
        zeroS = self.dare_zero_compare(A,B,Q,R,S)
        zeroS_new = self.dare_zero_compare(A,B,Q,R,S_new)

        for ai in zeroS:
            for aij in ai:
                self.assertAlmostEqual(aij, 0.0)
                raise AssertionError

### PR DESCRIPTION
**Pull request by ilayn**
This fixes #2251 and its tests per the discussion on the issue page. The doc string is updated.

### CODE REVIEW

**Comment by person142** [comment]
BTW you can check all the pep8 failures locally by installing pep8 (`pip install pep8`) and running `pep8 scipy` from the top-level scipy directory. It's looking like that will be the last thing you need for the Travis build to succeed.

**Comment by ilayn** [comment]
Bah, I guess I'm not cut for this. I've checked the files thrice with pep8 and it still fails.

**Comment by ilayn** [comment]
I've slept over it checked with some more examples and I think this PR is ready for review. Thanks .

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: I know it was like this in the original, but it would be better to put this in a references section and just cite it here. See:

https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#docstring-standard

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: Probably would fit better in the notes section.

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: Should this just define `H = [...]`, `J = [...]` instead of restating the equation? I guess you could also take out the first `H - \lambda J`.

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: I don't *think* other `linalg` routines do checks like this, so I wouldn't put them here. But I might be wrong.

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: Nitpick: no extra line here or between the `if`s.

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: Could you say more about the TODO? Are there places where this could fail?

File: scipy/linalg/tests/test_solvers.py
Comment (person142) [inline scipy/linalg/tests/test_solvers.py]: Safe to take out the "corrected for hermitian Q" bit--if someone can't see this diff they won't know what it means.

**Comment by person142** [comment]
Though the previous method doesn't always work, it still might be nice to see how the two compare speed-wise. (Especially on more sizable problems.) Benchmarks would go here:

https://github.com/scipy/scipy/blob/master/benchmarks/benchmarks/linalg.py

You can run them with `python runtests.py --bench` (you will need `asv`; `pip install asv`). An intro to writing them is here:

https://asv.readthedocs.io/en/latest/writing_benchmarks.html

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: These equations don't have solvability guarantees. Depending on the data matrices a solution might not exist. I don't know what SciPy way when no solution is found : return empty matrix or raise exception ?

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: If we don't do it, it will trip somewhere in the code with the cryptic broadcast exception. I think it is better to show why it fails such that the user immediately spots the mistake.

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: That's a domain specific formulation. I've moved all of those into Notes.

**Comment by ilayn** [comment]
Unfortunately, on a windows machine I can't do any of the build related stuff including the `runtests`. But since the previous one was not reliable a speedtest won't make too much of a difference. There are many extras to this implementation to make it comparable to matlab. But this version is at least consistent.

**Comment by person142** [comment]
> But since the previous one was not reliable a speedtest won't make too much of a difference.

I guess I worry that someone will have working code that suddenly slows down a lot. Breaking their case to make other cases work might not make them happy. If there's a significant speed difference it's possible that we might want to have both versions provided that for the old one there's a way to either

- document a reasonable subset of problems on which it will work or
- detect failures and raise an exception.

**Comment by person142** [comment]
Also it's possible to run benchmarks without `runtests.py` by doing something like `asv dev --bench <benchmark>` which won't trigger a build and only runs the benchmark once. (So take the results with a grain of salt.)

**Comment by person142** [comment]
It would be nice to include more DAREX examples in the tests.

**Comment by ilayn** [comment]
Indeed but then I have to include balancing algorithms which is basically copying everything from [my other implementation](https://github.com/ilayn/harold/blob/master/harold/harold.py#L6009-L6400) which doesn't seem that easy to convert to SciPy way without breaking some back-compatibility.

**Comment by person142** [comment]
> Indeed but then I have to include balancing algorithms

I don't see why? I had something like this in mind (only added a few examples):

https://github.com/person142/scipy/blob/ilayn/scipy/linalg/tests/test_solvers.py#L135

The tests will make some noise until https://github.com/scipy/scipy/pull/6582 is in.

(Not for now, but maybe in the future: adding balancing seems possible--you know your code better than me, but it seems like it would just require adding a new keyword argument, which is back-compatible.)

**Comment by ilayn** [comment]
Ah I might have been too quick to type. The main reason is that balancing is a really strange concept and based on tweaks that have been eluding me for some time. It's partially due to my Fortran skills are being next to none. 

I made the Riccati solvers work just to keep up with matlab in terms of accuracy but matlab is certainly better for now. I actually tried to ask around what might be the trick ( see [this thread on scicomp.SE](http://scicomp.stackexchange.com/questions/24937/matrix-balancing-algorithm) ) but so far no feedback. 

I have also checked the other solvers on `_solvers.py` and indeed I can be of help with them too. But my preference is first to iron out this buggy behavior and in the meantime if I can get things done, I'll write a major PR for both factoring out the balancing if I can. For relatively OK well-conditioned problems this PR is also sufficient as the examples in CAREX and DAREX are really really ill-conditioned thus not so relevant for the average user. 

Note that current equations are generally called the regular CARE and DARE. There is also two more versions of these namely GDARE and GCARE and the code I have on my package generalizes that (with extra matrices S and E comes into play but they are rarely used). 

Hence, if I may, I'd like to keep the embarassment minimal until I master the balancing ;) 

PS : An alternative route is to lay a post-newton iteration scheme such that it takes the whatever solution came out of the current solver and uses it as the initial condition to iteratively improve the solution. That's also something I have been working on and it feels like I'm getting closer to a finished product. Long story short, I will come back to these again in a more structured fashion with an optional keyword to turn that behavior on and off.

**Comment by ilayn** [comment]
By the way superb job on finding and adding the examples that fast. Great stuff!

**Comment by ilayn** [comment]
@person142 What would be the best way to proceed with this? Wait until we iron out everything and do it with one shot?  Or fix the wrong result with this patch and do another major PR for a general focus also modify continuous version and include the balancing for both ? 

Again, as a reminder the balancing is still not perfect. I'm still working on it.

**Comment by person142** [comment]
> Or fix the wrong result with this patch and do another major PR for a general focus

I like this one. I'm still in favor of putting in DAREX examples, however. Not all of them will fail (the only three I tried work fine), and they provide extra coverage that wasn't possible before (e.g. singular `r`). If some of them fail then we can mark them as known failures (and they can be a target which future pull requests are aiming for); I just want to test the boundaries of how far this patch takes us.

**Comment by ilayn** [comment]
@person142  Oh that I can tell. In the CAREX examples, there are two examples fail even after my balancing implementations (ex.6 and 12 in the matlab version). In DAREX I have to finish the conversion from matlab to Python but so far so good but I'm sure it will fail on a few suspicious looking ones. 

One major pain is the PEP8 and those matrices don't go well together, is there a way to ignore PEP8 temporarily of things or what is the preferred way of entering them? Tall Submatrices ?

**Comment by pv** [comment]
If the matrices are big, I would put them as data files (e.g. text files), under `scipy/linalg/tests/data`, and load via `mat = np.loadtxt(os.path.join(os.path.dirname(__file__), 'data', 'darex_1.dat')` etc.
If they are small, I would just follow pep8 format.

**Comment by ilayn** [comment]
`darex #12` is what I mean by having a relatively accurate but not acceptable in terms of absolute terms solution, as the solution should be `diag(1,1e12)` but gets approx. `diag(1,1e12 + 2.5e7)`.

**Comment by person142** [comment]
Ok, for the benefit of anyone looking at this: with the new set of tests the old function has 4 exceptions due to ill-conditioning in `r` and one failure. See e.g. here:

https://github.com/person142/scipy/tree/old-discrete-are

File: scipy/linalg/tests/test_solvers.py
Comment (person142) [inline scipy/linalg/tests/test_solvers.py]: I have some suggested changes for this test:

https://github.com/person142/scipy/blob/ilayn/scipy/linalg/tests/test_solvers.py#L136

It turns the failing tests into known failures so that they're more visible.

**Review by person142 — CHANGES_REQUESTED** [review]
Ok, I left some comments. They're mostly pretty small things; I think this is pretty close to being done.

File: scipy/linalg/_solvers.py
Comment (pv) [inline scipy/linalg/_solvers.py]: Needs `::` (two colons) at the end of the preceding text, ie. `... given by the block matrices::`

File: scipy/linalg/tests/test_solvers.py
Comment (person142) [inline scipy/linalg/tests/test_solvers.py]: For got to mention this: reference the original github issue here so it's easy to find later.

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: I didn't get your point on spacing

File: scipy/linalg/tests/test_solvers.py
Comment (ilayn) [inline scipy/linalg/tests/test_solvers.py]: Ah, forgot to include this...

File: scipy/linalg/_solvers.py
Comment (person142) [inline scipy/linalg/_solvers.py]: It's not a show-stopper, but I'd like related groups of if statements to not have any spaces between them (and maybe be converted to if/elif's) so that they cluster together visually. But this might just be a personal preference so I won't press the point.

File: scipy/linalg/tests/test_solvers.py
Comment (person142) [inline scipy/linalg/tests/test_solvers.py]: Ok so I got a little lazy putting in the causes for failure here; should probably at least be something like "unsatisfactory absolute error" if that's the cause.

Comment (ilayn) [inline scipy/linalg/tests/test_solvers.py]: No actually it is a complete failure to find a valid solution. The other two I marked it as bad accuracy

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: I thought you wanted a spacing between the ifs inside the for loop so I extrapolated from that, but I have no preference.

**Comment by ilayn** [comment]
This hopefully concludes the requested changes. Please let me know if there are additional items

**Review by person142 — APPROVED** [review]

**Comment by person142** [comment]
I'm happy (the CircleCI passed, not sure why it's still yellow), but I'll wait to get a second opinion from another developer.

File: scipy/linalg/_solvers.py
Comment (pv) [inline scipy/linalg/_solvers.py]: This will bomb if `u_sym` is real, because in-place operations on overlapping memory regions is undefined behavior.
For extra fun, you don't see it for small matrices due to buffering.

**Comment by pv** [comment]
Looks OK to me apart from that one cmment.

File: scipy/linalg/_solvers.py
Comment (ilayn) [inline scipy/linalg/_solvers.py]: Wow. That's really good to know. I'll fix it in a moment. Is there any place where I can read more about it?

**Comment by person142** [comment]
Also would be good to squash things.

File: scipy/linalg/_solvers.py
Comment (pv) [inline scipy/linalg/_solvers.py]: It's the issue about Numpy arrays and views to numpy arrays, eg `b = a[::2]; b[0] = 1` modifies also `a`. I think it's mentioned in most numpy tutorials. With in-place operations etc., you want to make sure the output operand is not a view of any kind of the input.

However, I think this caveat will go away in near future https://github.com/numpy/numpy/pull/8043. There are also long threads about this in the numpy issue tracker, but I'd recommend those only for entertainment purposes now that there's a consensus on how the issue will be addressed.

**Comment by pv** [comment]
@person142: there's an option to squash and merge in the github merge button

**Comment by ilayn** [comment]
I don't know what "squash" refers to actually.

**Comment by pv** [comment]
Thanks, merged!

**Comment by person142** [comment]
Great! Thanks for sticking with us @ilayn, and I look forward to your future improvements!

**Comment by ilayn** [comment]
Pleasure is mine.

**Pull request merged**
